checker/src/org/checkerframework/checker/genericeffects/GenericEffectVisitor.java

```diff
@@ -148,69 +138,958 @@ public Void visitMethod(MethodTree node, Void p) {
+            return super.visitMethodInvocation(node, p);
+        }
+        if(hasEnclosingVariable()) {
+            ExecutableElement varElt = TreeUtils.elementFromUse(node);
+            Class<? extends Annotation> varTargetEffect = atypeFactory.getDeclaredEffect(varElt);
+            checkEnclosingVariable(node,
+                    varTargetEffect,
+                    "call.invalid.super.effect",
+                    null);
+            return super.visitMethodInvocation(node, p);
+        }
+        return super.visitMethodInvocation(node, p);
     }
 
-    /*
-     * Method to check if the constructor call is made from a valid context
+    /**
+     * Method to check if the constructor call is made from a valid context.
+     *
+     * @param node New class tree node that is found during checking.
+     * @param p Void
+     * @return Void
      */
     @Override
     public Void visitNewClass(NewClassTree node, Void p) {
-        if (debugSpew) {
-            System.err.println(
-                    "For constructor " + node + " in " + currentMethods.peek().getName());
+        if(hasEnclosingMethod()) {
+            ExecutableElement methodElt = TreeUtils.elementFromUse(node);
+            Class<? extends Annotation> targetEffect = atypeFactory.getDeclaredEffect(methodElt);
+            checkEnclosingMethod(node,
+                    targetEffect,
+                    "constructor.call.invalid",
+                    null);
+            return super.visitNewClass(node, p);
         }
+        if(hasEnclosingVariable()) {
+            ExecutableElement varElt = TreeUtils.elementFromUse(node);
+            Class<? extends Annotation> varTargetEffect = atypeFactory.getDeclaredEffect(varElt);
+            checkEnclosingVariable(node,
+                    varTargetEffect,
+                    "constructor.call.invalid",
+                    null);
+            return super.visitNewClass(node, p);
+        }
+        return super.visitNewClass(node, p);
+    }
 
-        // Target method annotations
-        ExecutableElement methodElt = TreeUtils.elementFromUse(node);
-        if (debugSpew) {
-            System.err.println("methodElt found");
+    /**
+     * The methods below this comment follow the same format. Each method is a different visit method
+     * for a different kind of tree node. Using the extensions class the developer can activate specific
+     * visitor methods depending on what they want to check.
+     *
+     * The methods work by first checking if the node being checked is enclosed by a method. If it is then
+     * the method obtains the effect of the node and checks it against the method's effect. If the node is not
+     * enclosed by a method, then it checks at the variable level against the class annotation.
+     *
+     * @param node Specific tree node that is to be checked.
+     * @param p Void
+     * @return Void
+     */
+    @Override
+    public Void visitArrayAccess(ArrayAccessTree node, Void p) {
```

Comment:
This method, and the others following the same template, all need a couple fixes.

1. The call to ```super.visitArrayAccess(node, p)``` shouldn't need to be repeated --- it should just be the last line of the method.
2. Right now, if an extension *doesn't* do a certain check, the checker won't recur deeper into that structure!  For example, if a given extension doesn't perform an array access check, then control jumps straight to ```return null```, without visiting any subexpressions.  This will result in missing lots of errors, and needs to be fixed in all of the extension points below as well.
3. I'm starting to think the ```checkEnclosingMethod``` is over-engineered.  At any extension point, we're always asking the extension what a possible error might be, *before* we know if there's actually a problem to report.  It seems like it might be preferable to have each extension point do its own inline check (either explicitly, or via a helper method) to determine first whether there's any issue at all (if the targetEffect is invalid for this context), and *only if there is* ask for error or warning strings.